test/studies/bale/toposort/toposort.chpl

```diff
@@ -103,115 +147,558 @@ class Vector {
     this.size += 1;
   }
 
-  proc push_back( values : Vector(eltType) ){
+  /*
+  Push back all values from another Vector.
+  :arg values: Vector of values to copied and pushed onto end of this Vector.
+  :type values: unmanaged Vector(eltType).
+  */
+  proc push_back( values : unmanaged Vector(eltType) ){
     while this.size + values.size > this.listDomain.size {
       this.listDomain = {1..#( this.factor * this.listDomain.size )};
     }
     this.listArray[size+1..#values.size] = values.listArray;
     this.size += values.size;
   }
 
+  /*
+  Push back all values from a list.
+  :arg values: list of values to copied and pushed onto end of this Vector.
+  :type values: list(eltType).
+  */
   proc push_back( values : list(eltType) ){
     for value in values {
       this.push_back( value );
     }
   }
 
+  /*
+  Empty queue.
+  */
   proc clear( ){
     this.size = 0;
   }
 
+  /*
+  Compact underlying array to minimal size.
+  .. note: This leaves extra no extra capacity for new values.
+    Will cause a reallocation after a push-back.
+  */
   proc compact(){
     this.listDomain = {1..#this.size};
   }
+
+  /*
+  Yield all values from array.
+  :ytype: eltType
+  */
+  iter these(){
+    for i in 1..#size {
+      yield this.listArray[i];
+    }
+  }
+
+  proc writeThis( f ){
+    f <~> this.listArray[this.listDomain.low..#this.size];
+  }
+}
+
+/*
+Queue for effecient shared-memory parallel work queueing and distributing using
+a chuncked loop wait method.
+*/
+class ParallelWorkQueueChunkedLoopWait {
+  type eltType;
+  type lockType;
+
+  const maxTasks : uint;
+  const startTaskID : uint = 0;
+  const taskDomain : domain(1,uint);
+
+  var locks : [taskDomain] unmanaged lockType;
```

Comment:
The locks don't necessarily protect the Vector as a whole, rather they protect access to the pointer/reference to it. That seems like a poor explanation/reason for them being separated, but really the key locking happens [in the `these` iterator](https://github.com/chapel-lang/chapel/pull/10829/files/20ca8517fa8388c209cf3a2243544e250dc1d6fc?utf8=%E2%9C%93&diff=unified#diff-7fcece9fe9782ce8d199d77fb1f17e1aR317), where a task's Vector is quickly (hopefully) swapped and it can be worked on by the task safely with no internal locks.

Im not sure how that swap would work with an vector-internal lock.

Reply:
Either way, I'd recommend adding padding to these `locks` to avoid false sharing (#10771)

Reply:
This has now been addressed by the [`ParallelWorkQueueChunkedLoopWaitPadding`](https://github.com/chapel-lang/chapel/pull/10829/files#diff-7fcece9fe9782ce8d199d77fb1f17e1aR382), and can used when run with `--implementation ParallelChunkedLoopWaitPadding`.

And it absolutely works. In one case I get a 22% increase in rows/second.

Reply:
After examining performance across a wide range of padding amounts, I've come to the conclusion that it's probably not the padding, but maybe the conjoining of the lock and the vector in tuple that helps here.